### PR TITLE
[FLINK-15904][connectors/kafka] Make Kafka Consumer work with activated "disableGenericTypes()"

### PR DESCRIPTION
**Pull request by JTaky**
## What is the purpose of the change

Currently, class KafkaTopic KafkaTopicPartition uses GenericTypeInformation during the serialization in the unionOffsetStates inside the KafkaConsumer. Which prevents any flink application to `disableGenericTypes`.
We can still continue to use compatible KryoSerialization, but by specifying serialization explicitly we can avoid check on generic types.

## Brief change log

- Replace TypeHint in the state initialization by explicit TupleSerializer which takes KryoSerializer as the parameter.

## Verifying this change

I suppose that test FlinkKafkaConsumerBaseMigrationTest#testRestoreFailsWithNonEmptyPreFlink13StatesIfDiscoveryEnabled will test it against several states savepoints from the previous version.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: don't know
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? No

### CODE REVIEW

**Comment by JTaky** [comment]
@flinkbot run travis

**Comment by aljoscha** [comment]
@JTaky I like how you put the code in a separate method! I pushed a commit that simplifies the test quite a bit. Could you check if that's ok for you?

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java [hunk 1]
```diff
@@ -675,39 +672,35 @@ private void setupConsumer(FlinkKafkaConsumerBase<String> consumer) throws Excep
...
+		TupleSerializer<Tuple2<KafkaTopicPartition, Long>> kafkaConsumerSerializer =
+				FlinkKafkaConsumerBase.createStateSerializer(executionConfig);
+		Tuple2<KafkaTopicPartition, Long> actualTuple =
+				InstantiationUtil.deserializeFromByteArray(kafkaConsumerSerializer, bytes);
 
 		Assert.assertEquals(
-			"Explicit KryoSerializer for Tuple2<KafkaTopicPartition, Long> in not compatible with TypeHint based serializer",
-			expectedTuple,
-			actualTuple
+				"New Serializer is not compatible with previous method of creating Serializer using TypeHint.",
```
Comment (JTaky) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: imho new it a very relative term and gives no information once this test fails.
What about: "Explicit Serializer is not compatible with previous method of creating Serializer using TypeHint"

**Comment by JTaky** [comment]
> @JTaky I like how you put the code in a separate method! I pushed a commit that simplifies the test quite a bit. Could you check if that's ok for you?

Thanks for the change!
Have only a single optional comment about UT error message.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java (hunk 1 above)
Comment (aljoscha) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Yes, will change that. 👌

**Comment by aljoscha** [comment]
Thanks for working on this! 😃 I merged.

**Pull request closed without merging**